sxg_rs/src/crypto.rs

```diff
@@ -0,0 +1,165 @@
+        let octets = der_parser::parse_ber(der)?
+            .1
+            .as_slice()
+            .map_err(|e| Error::new(e).context("Expecting ECPoint to be an OCTET STRING"))?;
+        const KEY_SIZE: usize = 32; // Both X and Y of the EC Point are 32 bytes (256 bit).
+        if octets.len() != 1 + KEY_SIZE * 2 {
+            return Err(Error::msg(format!("Expecting ECPoint to contain 1 octet of uncompression flag and {}*2 octets of point coordinates", KEY_SIZE)));
+        }
+        match octets[0] {
+            0x04 => (),
+            0x03 => {
+                return Err(Error::msg(
+                    "We don't support ECPoint in compressed form, please use uncompressed form.",
+                ))
+            }
+            _ => return Err(Error::msg("Invalid ECPoint form")),
+        };
+        let mut x = octets[1..].to_vec();
+        let y = x.split_off(KEY_SIZE);
+        Ok(EcPublicKey {
+            kty: "EC".to_string(),
+            crv: "P-256".to_string(),
+            x,
+            y,
+        })
+    }
+}
+
+impl EcPrivateKey {
+    pub fn from_sec1_der(der: &[u8]) -> Result<Self> {
+        let ec_private_key = der_parser::parse_ber(der)?.1;
+        // https://datatracker.ietf.org/doc/html/rfc5915#section-3
+        //   ECPrivateKey ::= SEQUENCE {
+        //     version        INTEGER { ecPrivkeyVer1(1) } (ecPrivkeyVer1),
+        //     privateKey     OCTET STRING,
+        //     parameters [0] ECParameters {{ NamedCurve }} OPTIONAL,
+        //     publicKey  [1] BIT STRING OPTIONAL
+        //   }
+        let d = ec_private_key
+            .as_sequence()
+            .map_err(|e| Error::new(e).context("Expecting ECPrivateKey to be a SEQUENCE"))?
+            .get(1)
+            .ok_or_else(|| Error::msg("Expecting ECPrivateKey to contain at least 2 items"))?;
+        let d = d
+            .as_slice()
+            .map_err(|e| Error::new(e).context("Expecting privateKey to be an OCTET STRING"))?
+            .to_vec();
+        let public_key = EcPublicKey::from_sec1_der(
+            ec_private_key
+                .as_sequence()?
+                .get(3)
+                .ok_or_else(|| Error::msg("Expecting ECPrivateKey to contain at least 4 items"))?
+                .as_slice()?,
+        )?;
+        Ok(EcPrivateKey { d, public_key })
+    }
+    pub fn from_sec1_pem(pem: &str) -> Result<Self> {
+        let der = get_der_from_pem(pem, "EC PRIVATE KEY")?;
+        Self::from_sec1_der(&der)
+    }
+    #[cfg(feature = "rust_signer")]
+    pub fn create_signer(&self) -> Result<RustSigner> {
+        RustSigner::new(&self.d)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    fn assert_bytes_eq_hex(bytes: Vec<u8>, hex: &str) {
+        let bytes: Vec<String> = bytes.iter().map(|x| format!("{:02x}", x)).collect();
+        let mut hex = hex.to_string();
+        hex.retain(|c| !c.is_whitespace());
+        assert_eq!(bytes.join(":"), hex);
+    }
+    #[test]
+    fn it_works() {
```

Comment:
Could you add a test where it's supposed to fail? Thanks.

Reply:
Done.